### ISSUE TITLE
pandas results serializer proposal 

### ISSUE THREAD

**Issue by bobcolner** [PrefectHQ/prefect#2917]
Pandas dataframes are very common in python data workflows, often saved to disk in efficient binary formats. 

I'm sharing  parquet and feather serializers that I am using and might also benefit others. 

I'm not sure where or even if these pandas specific serializer fits into prefect core. What does are the teams thoughts? If we go down this route we will probably want to add CSV as well.
 

```
from io import BytesIO
import pandas as pd
import pyarrow.feather as pf
from prefect.engine.serializers import Serializer


class FeatherSerializer(Serializer):

    def serialize(self, value: pd.DataFrame) -> bytes:
        # transform a Python object into bytes        
        bytes_buffer = BytesIO()
        pf.write_feather(
            df=value,
            dest=bytes_buffer,
            version=2,
        )
        return bytes_buffer.getvalue()

    def deserialize(self, value:bytes) -> pd.DataFrame:
        # recover a Python object from bytes
        df_bytes_io = BytesIO(value)
        return pd.read_feather(df_bytes_io)


class ParquetSerializer(Serializer):

    def serialize(self, value: pd.DataFrame) -> bytes:
        # transform a Python object into bytes
        bytes_buffer = BytesIO()
        value.to_parquet(
            path=bytes_buffer,
            index=False
        )
        return bytes_buffer.getvalue()

    def deserialize(self, value:bytes) -> pd.DataFrame:
        # recover a Python object from bytes        
        df_bytes_io = BytesIO(value)
        return pd.read_parquet(df_bytes_io)
```

**Comment by jcrist** [comment]
Nice work @bobcolner. My personal thoughts are that we should experiment with pandas specific serializers outside of prefect for a bit, just to see how they feel. Others may feel differently though.

Also, cc'ing @AndrewRook for thoughts, who's been thinking about this for a while (and submitted #2115, which kicked off the serializers in results work).

**Comment by AndrewRook** [comment]
I may be rapidly losing my authority on the topic as I've not had much time to spend on this recently, but my personal perspective is that the best way to handle pandas serializing is to provide users with a single entry point — instead of having a `CSVSerializer`, `ParquetSerializer`, and so on, you provide a `PandasSerializer` that takes in how you want to serialize as an argument. I was thinking the options would be only the things that pandas supports (ie only stuff with a `DataFrame.to_[thing]` method), although I'm not sure where that would leave the feather support since in your example the serializer doesn't use a DataFrame method.

**Comment by bobcolner** [comment]
Yep, one  issue is that the pandas to_feather()

**Comment by bobcolner** [comment]
sorry, I fat finger closed this issue.

**Comment by AndrewRook** [comment]
FYI I started working on an implementation of this and rapidly ran into a longstanding pandas bug with writing to binary handles. There is an open PR (pandas-dev/pandas#35129) that is working to resolve this issue, at least for CSVs, but I'm not sure if it's worth waiting for that to get into a release (and it also doesn't help all the folks who aren't on bleeding edge versions of pandas, although I'm not sure how much we really care there). 

Workarounds I've seen suggested or can think of myself seem kludgey (e.g. writing to a string buffer and then bytes-encoding), and I'm not experienced enough with the ins and outs of file I/O to really be able to evaluate them for potential pitfalls. Thoughts anyone?

**Comment by lauralorenz** [comment]
Hi all, I want to loop in work and conversation happening in https://github.com/PrefectHQ/prefect/pull/2963 which is related to the pandas serializer discussion! It's more on the "one serializer to rule them all" implementation side.

**Comment by lauralorenz** [comment]
Update on the last comment: Looks like we are pretty bought in on 'one serializer to rule them all', internally and among the other folks talking about the open forks/PRs on that linked PR. As of now I think we are going to go that direction. 

Fundamentally I think this means in a way we 'reject' the specific implementation proposal of OP in the sense of separate serializer classes, but the overall concept of getting pandas-derived serializers into core that can produce a deserialized DataFrame from varying serialization formats is a go 👍

**Comment by jcrist** [comment]
Fixed by #3020, closing.
